data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/CatalogController.java

```diff
@@ -0,0 +1,119 @@
+import org.eclipse.edc.catalog.spi.Catalog;
+import org.eclipse.edc.jsonld.transformer.JsonLdTransformerRegistry;
+import org.eclipse.edc.protocol.dsp.catalog.spi.types.CatalogRequestMessage;
+import org.eclipse.edc.spi.EdcException;
+import org.eclipse.edc.spi.iam.IdentityService;
+import org.eclipse.edc.spi.iam.TokenRepresentation;
+import org.eclipse.edc.web.spi.exception.AuthenticationFailedException;
+import org.eclipse.edc.web.spi.exception.InvalidRequestException;
+
+import java.util.ArrayList;
+import java.util.Map;
+
+import static java.lang.String.format;
+import static java.lang.String.join;
+import static org.eclipse.edc.jsonld.util.JsonLdUtil.compact;
+import static org.eclipse.edc.jsonld.util.JsonLdUtil.expand;
+import static org.eclipse.edc.protocol.dsp.catalog.spi.CatalogApiPaths.BASE_PATH;
+import static org.eclipse.edc.protocol.dsp.catalog.spi.CatalogApiPaths.CATALOG_REQUEST;
+import static org.eclipse.edc.protocol.dsp.catalog.transform.DspCatalogPropertyAndTypeNames.DSPACE_PREFIX;
+import static org.eclipse.edc.protocol.dsp.catalog.transform.DspCatalogPropertyAndTypeNames.DSPACE_SCHEMA;
+import static org.eclipse.edc.protocol.dsp.transform.transformer.Namespaces.DCAT_PREFIX;
+import static org.eclipse.edc.protocol.dsp.transform.transformer.Namespaces.DCAT_SCHEMA;
+import static org.eclipse.edc.protocol.dsp.transform.transformer.Namespaces.DCT_PREFIX;
+import static org.eclipse.edc.protocol.dsp.transform.transformer.Namespaces.DCT_SCHEMA;
+import static org.eclipse.edc.protocol.dsp.transform.transformer.Namespaces.ODRL_PREFIX;
+import static org.eclipse.edc.protocol.dsp.transform.transformer.Namespaces.ODRL_SCHEMA;
+
+/**
+ * Provides the HTTP endpoint for receiving catalog requests.
+ */
+@Consumes({MediaType.APPLICATION_JSON})
+@Produces({MediaType.APPLICATION_JSON})
+@Path(BASE_PATH)
+public class CatalogController {
+    
+    private ObjectMapper mapper;
+    private IdentityService identityService;
+    private JsonLdTransformerRegistry transformerRegistry;
+    private String dspCallbackAddress;
+    
+    public CatalogController(ObjectMapper mapper, IdentityService identityService,
+                             JsonLdTransformerRegistry transformerRegistry, String dspCallbackAddress) {
+        this.mapper = mapper;
+        this.identityService = identityService;
+        this.transformerRegistry = transformerRegistry;
+        this.dspCallbackAddress = dspCallbackAddress;
+    }
+    
+    @POST
+    @Path(CATALOG_REQUEST)
+    public Map<String, Object> getCatalog(JsonObject jsonObject, @HeaderParam(HttpHeaders.AUTHORIZATION) String token) {
+        var expanded = expand(jsonObject).getJsonObject(0); //expanding returns a JsonArray of size 1
+        var messageResult = transformerRegistry.transform(expanded, CatalogRequestMessage.class);
+        if (messageResult.failed()) {
+            throw new InvalidRequestException("Request body was malformed.");
+        }
+    
+        checkAuthToken(token);
```

Comment:
this could be done before the transformation

Reply:
Not sure about this. Is the token being verified or its structure checked? If it is the former, I think that should be done outside the transformer (they only convert from one structure to another), otherwise, putting it in the transformed would make sense.

Reply:
It's about token verification, but I understood the comment as simply moving the checking of the token to be the first operation in the controller method. So if the token is not valid, the transformation of the request body will not be performed. And I think that makes sense.